### ISSUE TITLE
Secure backup setup

### ISSUE THREAD

**Issue by Yannik** [borgbackup/borg#3608]
Hello everyone,

first, I want to state that I have read various issues (like #900) which were somehow related to this.

A secure backup setup should IMO have the following properties:

## No single point of failure
If the device being backup up is compromised, the backup needs to stay in tact. If the backup destination is compromised, the host being backup needs to stay in tact.

## The device being backed up should not be able to delete its own backup
Otherwise an attacker could encrypt your backups (ransomware), or just delete them. 
As a result, any type of "push" backup is not viable, for the exception of the target being append-only (does something like this exist)?
Therefore, I'll now focus on pull-based backup setups.

## The host pulling the backup should only have read-only access, restricted to the directories that are being backed up
Obviously, if the host pulling the backups has access to the host being backed up, a compromised backup-pulling host can just overtake all the hosts that it backs up. Therefore, it needs to have strict read-only access. However, this needs to be restricted to the directories that are being backed up. Otherwise, the backed up ssh key will grant immediate access.

Have these issues been thought about while creating borg backup? I'm currently rolling a backup solution based on rsnapshot which takes care of these issues (see https://github.com/Yannik/ansible-role-rsnapshot-backup-host, https://github.com/Yannik/ansible-role-rsnapshot-remote-host), but I would like to move to a solution which allows for differential file backups.

**Comment by ThomasWaldmann** [comment]
Borg has push mode, but no built-in pull mode. Assumptions about pull mode being more safe than push are based on high security assumptions about the backup repo server, which might be correct or just a nice illusion.

Borg does have append-only mode, which avoids that a client can physically delete data.
It is a rather low level hack and not very comfortable to use (in case of a hack), but it does what it says.
The problem of actually having to know when you got hacked and not kill your backups when you still need them is a very fundamental one and not related to borg. Same goes for knowing whether your backups (still) contain what you expect them to contain.

Borg client being able to delete backups does not imply ransomware being able to encrypt your backup repo. This is only (as always) the case if you allow fs access to the repo (e.g. local directory, samba, nfs, sshfs, ...), but does not apply to borg client/server mode (ssh: repo url).

Good idea about the pull-mode that is restricted to read-only and only to the to-be-backed-up directories.
But as long as we have no pull mode, this is just a enhancement request for a non-existing feature.

rsnapshot has its own issues btw. due to its low tech approach, guess you can find them on the internet yourself.

A relatively safe borg setup using the currently existing features would:
- use client/server, authorized_keys on repo server with restrictions (ssh-key only, append-only, only access to own repo, ...)
- alternatively to borg append-only you could also do snapshots on fs level on the repo server, e.g. with zfs
- either never prune (safe, but space consuming) or make sure your repo/archives are in good state, use separate ssh key and no append-only for prune
- as it is push, you don't have the issue of the repo server being root on any backup client

**Comment by Yannik** [comment]
Actually I thought about the snapshot approach. Even though I don't have experience with zfs, my previous experiences with snapshots is, that it is a right mess to free up space as well. IMO snapshots duplicate the features that should be provided by differential backup software. Otherwise (this is a simplification!) I could just rsync and create a snapshot every night.

The approach you suggest, using append-only, has the issues we both already mentioned:
  - Either never prune (not really viable)
  - or ensure that repo/archives are in a good state, which, depending on the level of sophistication of the attacker (we are talking about penetrating a specific user account on a linux system, which should not be too easy to begin with), can be very hard, and will be, in any case, time consuming (do this for X repos)

Neither the former nor the latter issue really make borg a viable backup solution for production. This is not a rant, I would really like to use it, but it appears to me, that it is not usable with security in mind right now.

Is this a concious decision? I see that there a quite a few bountys posted by borgbackup on various (imo) not-really-important things, but there seems to be no traction on this topic (and related bugs  #2251, #1772) for a long time.

**Comment by ThomasWaldmann** [comment]
It's a matter of priorities:

- 1.0: fixes and essential, small features
- 1.1: security improvements, a few bigger features, tuning
- 1.2: better crypto and multithreading

The client/server code is complex (esp. its async parts), hard to test and issues with it hard to debug, so the prio with that was also rather to stabilize/fix it and make it easier to add new rpc methods / params without breaking old clients than to vastly extend or even duplicate it to implement pull mode.

**Comment by Yannik** [comment]
Hey @ThomasWaldmann, do you see any improvements with regards to these issues in the coming months?

**Comment by ThomasWaldmann** [comment]
The coming months will focus on 1.2, thus: no.

**Comment by DFoIUtN02r** [comment]
@ThomasWaldmann

>Assumptions about pull mode being more safe than push are based on high security assumptions about the backup repo server, which might be correct or just a nice illusion.

I don't think this is an illusion, but a basic imperative for designing any backup setup. Of course, it's up to the administrator to ensure that this is indeed the case. But if your servers are more likely to be hacked that your clients, that's a big issue.

The offline backup is also just another level of security, requiring manual intervention. That way you have three escelating levels of security:
* clients: protected by (often insecure) passwords that users frequently type every day (vulnerable to phishing and logging)
* backup-server: protected by the admin's (hopefully secure) password and any additional keys
* offline backup: protected by physical access to the storage and any additional encryption keys.

**Comment by Yannik** [comment]
@ThomasWaldmann I would like to bump this issue, as unfortunately, rsnapshot, which I've so far used for secure pull-backups, is being removed from debian in the next debian release (https://github.com/rsnapshot/rsnapshot/issues/279)
I would very much like to switch to borg, but I'm really hard pressed about doing that, with the basic backup security parameters outlined in my OP not being possible :(

**Comment by ThomasWaldmann** [comment]
IIRC we have pull mode in our docs meanwhile and #2125 might help with the read-only part.

**Comment by jonathanmmm** [comment]
Hi,
just thinking how to make borgbackup secure. Would it work, if the backup server has LVM/BTRFS file system an I do regular snapshots of the folder where all clients save their data? That way I could just keep append-only off and let clients do pruning, but LVM/BTRFS are actually backing up.

So in an event, that a client is corrupted and deletes all his backups or gets his file encrypted and pushes them into the repo and prunes old backups. I could just mount/look at an old snapshot, before this corruption and be able to restore from there, right?

This is a way, to ensure the backups are safe, at this stage of borg backups release. Of course I have to trust the backup server not to delete the snapshots, but that also is the same with the regular backups. I would just have to create a script that just has like a retention policy, like deleting LVM/BTRFS snapshots older than a month/year whatever would seem a good compromise.
Question is of course, how much and how fast could the disk of the backup server fill up, as every file that changes a bit, creates complete new blocks on the server (as they are encrypted).

But would that be still a viable and safe solution, until append-only or so forbids prune/delete commands?

**Comment by fatgoosee** [comment]
Hi,
just as an idea. How about letting the backup server handle prune on a schedule or condition and letting clients push their backups with append only?
I'm sure there is a way to script this.

**Comment by ThomasWaldmann** [comment]
@fatgoosee One of the borg threat model assumptions is that the backup server is not fully trusted. That's why there is usually no borg key on the backup server to decrypt data.

But to prune archives, you would need that, because archives and their metadata is encrypted.

**Comment by ThomasWaldmann** [comment]
borg 1.x has append-only mode which can be enforced server-side by a ssh forced command. It is a bit of a hack, but is documented and works.

borg2 (master branch) has more fine-grained permissions, also enforceable server-side.

Additionally to all that, admin could use snapshots on backup server fs or LV.

About pull mode there are other issues and pull requests open, but they are rarely used and definitely need more testing. For example, see #9459.
